Add RPO to CleanRL

masud99r opened the pull request:
## Description
<!--- Provide a general summary of your changes in here-->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [ ] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithm variants or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm variant.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format).
    - [ ] I have added links to the tracked experiments.
    - [ ] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have updated the tests accordingly (if applicable).

vwxyzjn commented:
Hey btw I noticed you ran into some issues with `AttributeError: 'GLFWContext' object has no attribute '_context'`. You just need to install some mujoco dependencies for rendering. https://github.com/vwxyzjn/cleanrl/blob/b558b2b48326d8bb8da7fa853914576ae4610f53/.github/workflows/tests.yaml#L154-L156

See https://github.com/deepmind/dm_control#rendering

vwxyzjn commented on cleanrl/rpo_continuous_action.py:
```diff
@@ -0,0 +1,332 @@
+        action_logstd = self.actor_logstd.expand_as(action_mean)
+        action_std = torch.exp(action_logstd)
+        probs = Normal(action_mean, action_std)
+        if action is None:
+            action = probs.sample()
+        return action, probs.log_prob(action).sum(1), probs.entropy().sum(1), self.critic(x)
```
A minor nit: you should commit your changes first before running the benchmark experiments. This way, the benchmarked experiments can be tracked to a commit :)

masud99r replied:
Thank you for the pointers and suggestions. I just committed the changes.

masud99r replied:
For the Mujoco dependency: Is there a `pip` version of the libraries that I can use? I am in a cluster machine where I do not have `sudo` access.

masud99r replied:
Unfortunately, I do not have Docker access in the cluster machine.

`poetry install` was fine, and I managed to run `Pendulum-v1` gym (Gymnasium) and recorded the result on openrlbenchmark. However, for some reason, not all my experiments run successfully. Out of 10 runs for the pendulum, only one progresses to about halfway. It seems the issue is with the threading mechanism in the benchmark file, as processes are getting killed by the cluster. I am planning to move my implementation to a local machine to where I have full access.

vwxyzjn replied:
That’s weird… I don’t know what happened

vwxyzjn replied:
Maybe try running the benchmark utility With -- workers 1

masud99r replied:
I figured a memory issue was killing the processes (out of memory). Not sure if the Pendulum-v1 (Gymnasium) environment requires more memory than the OpenAI Gym version. I have run the same environments with Gym, which runs fine (though with an older CleanRL version).

masud99r commented:
> Hey btw I noticed you ran into some issues with `AttributeError: 'GLFWContext' object has no attribute '_context'`. You just need to install some mujoco dependencies for rendering.
> 
> https://github.com/vwxyzjn/cleanrl/blob/b558b2b48326d8bb8da7fa853914576ae4610f53/.github/workflows/tests.yaml#L154-L156
> 
> See https://github.com/deepmind/dm_control#rendering. I added a bit of docs to explain this issue.

I have tried these commands on my local machine, but the issue persists. I just wanted to give an update.

vwxyzjn commented:
It might be the case you are using hardware-accelerated rendering. Maybe look up https://github.com/deepmind/dm_control#rendering and https://github.com/deepmind/dm_control/issues/136

masud99r commented:
Hi, did you get around this issue? https://github.com/Farama-Foundation/Shimmy/issues/15#issue-1446206487
I am facing a similar issue with dm_control.

vwxyzjn commented:
> Hi @vwxyzjn , did you get around this issue? [Farama-Foundation/Shimmy#15 (comment)](https://github.com/Farama-Foundation/Shimmy/issues/15#issue-1446206487) I am facing a similar issue with dm_control.

Actually yes, I can reproduce this error on my end. For now try running the code without `--capture-video`.

masud99r commented:
Here is the result comparison between PPO and RPO. These dm_control environments are used in the RPO paper. The RPO paper: https://arxiv.org/pdf/2212.07536.pdf.

PPO and RPO run for 8M timesteps, and results are computed over 10 random seeds.

|                                | ppo_continuous_action ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:-------------------------------|:----------------------------------------------------------------|:----------------------------------------------|
| dm_control/humanoid-stand-v0   | 21.41 ± 30.31                                                   | 140.84 ± 57.21                                |
| dm_control/humanoid-run-v0     | 5.47 ± 9.32                                                     | 23.82 ± 19.57                                 |
| dm_control/hopper-hop-v0       | 6.33 ± 18.58                                                    | 62.46 ± 88.61                                 |
| dm_control/quadruped-walk-v0   | 241.46 ± 92.15                                                  | 610.52 ± 215.04                               |
| dm_control/quadruped-run-v0    | 172.81 ± 40.87                                                  | 370.70 ± 119.32                               |
| dm_control/quadruped-escape-v0 | 22.78 ± 11.22                                                   | 67.82 ± 28.01                                 |
| dm_control/walker-stand-v0     | 451.71 ± 220.04                                                 | 724.61 ± 158.48                               |
| dm_control/walker-run-v0       | 124.14 ± 88.88                                                  | 389.88 ± 126.93                               |
| dm_control/fish-swim-v0        | 83.03 ± 8.93                                                    | 130.44 ± 51.66                                |

The rest of the dm_control experiments are running and will be added later.

vwxyzjn commented:
This is perfect. Looking forward to seeing the rest of the results. A quick tip is that instead of creating a separate `ppo_continuous_action_slurm` file, you could do `python ppo_continuous_action.py --exp-name ppo_continuous_action_8M --total-timesteps 8000000`

Feel free to prepare the documentation in `/docs/rl-algorithms/rpo.md`. https://github.com/vwxyzjn/cleanrl/blob/master/docs/rl-algorithms/ppg.md might be a good example.

masud99r commented:
Thanks for the tips. I was too worried not to mess up with existing experiments with the name `ppo_continuous_action`. Using the arg `--exp-name` is a better option.
Additionally, I wanted to direct the experiment's files (wandb, tensorboard, video, etc.) to a separate storage. This is not to overflow my workspace where I have code, as my workspace has a small storage budget. Thus I edited the `ppo_continuous_action.py` to make `ppo_continuous_action_slurm.py`. Is there a better way of doing that? Is there any arg that I can use?

As of the final merging, I will only keep the `rpo_continuous_action.py` and remove the rest of the files I used to run the jobs, including `ppo_continuous_action_slurm.py`. The code will still be available in the wandb experiments file (openrlbenchmark). Is that fine? Or let me know if there is a better way.

vwxyzjn commented:
> Thanks for the tips. I was too worried not to mess up with existing experiments with the name ppo_continuous_action.

This used to be a concern but not anymore since we are now doing filter-based workflow.

> Additionally, I wanted to direct the experiment's files (wandb, tensorboard, video, etc.) to a separate storage. This is not to overflow my workspace where I have code, as my workspace has a small storage budget. Thus I edited the

`--exp-name` should direct files to folders prefixed with `exp_name`, so it should have identical effects as creating a `ppo_continuous_action_slurm.py`. One other thing you might leverage is you could call the file from any directory and it will save the tensorboard and video files at that directory. For example, calling `python cleanrl/ppo.py` vs `cd cleanrl && python ppo.py` will have the storage files in different directories.

https://github.com/vwxyzjn/cleanrl/blob/2dd73af4a8f8c02de3d75f0f7f6e692dcf697273/cleanrl/ppo_continuous_action.py#L140-L153

> As of the final merging, I will only keep the rpo_continuous_action.py and remove the rest of the files I used to run the jobs, including ppo_continuous_action_slurm.py. The code will still be available in the wandb experiments file (openrlbenchmark). Is that fine? Or let me know if there is a better way.

There is no need to delete those experiments because the last benchmarked experiments are tagged with `latest`, which will be used for querying. So having these `ppo_continuous_action_slurm.py` would not affect any existing workflows. I would say that you should remove `failed` and `crashed` experiments.

masud99r commented:
> There is no need to delete those experiments because the last benchmarked experiments are tagged with `latest`, which will be used for querying. So having these `ppo_continuous_action_slurm.py` would not affect any existing workflows. I would say that you should remove `failed` and `crashed` experiments.

I am thinking about this PR. I have created some additional files (`run_rpo.sh`, `ppo_continuous_action_slurm.py`, etc.) to facilitate running the experiments. My understanding is that this PR will add two files to the `vwxyzjn/cleanrl`: `rpo_continuous_action.py`, and `rpo.sh` for the benchmarking.

vwxyzjn commented:
>  My understanding is that this PR will add two files to the vwxyzjn/cleanrl: rpo_continuous_action.py, and rpo.sh for the benchmarking.

That is correct. To keep track of the `ppo_continuous_action_slurm.py` experiments, I would say update the `ppo.sh` to add lines such as 
```
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
     --env-ids dm_control/acrobot-swingup-v0 dm_control/acrobot-swingup_sparse-v0 dm_control/ball_in_cup-catch-v0 dm_control/cartpole-balance-v0 dm_control/cartpole-balance_sparse-v0 dm_control/cartpole-swingup-v0 dm_control/cartpole-swingup_sparse-v0 dm_control/cartpole-two_poles-v0 dm_control/cartpole-three_poles-v0 dm_control/cheetah-run-v0 dm_control/dog-stand-v0 dm_control/dog-walk-v0 dm_control/dog-trot-v0 dm_control/dog-run-v0 dm_control/dog-fetch-v0 dm_control/finger-spin-v0 dm_control/finger-turn_easy-v0 dm_control/finger-turn_hard-v0 dm_control/fish-upright-v0 dm_control/fish-swim-v0 dm_control/hopper-stand-v0 dm_control/hopper-hop-v0 dm_control/humanoid-stand-v0 dm_control/humanoid-walk-v0 dm_control/humanoid-run-v0 dm_control/humanoid-run_pure_state-v0 dm_control/humanoid_CMU-stand-v0 dm_control/humanoid_CMU-run-v0 dm_control/lqr-lqr_2_1-v0 dm_control/lqr-lqr_6_2-v0 dm_control/manipulator-bring_ball-v0 dm_control/manipulator-bring_peg-v0 dm_control/manipulator-insert_ball-v0 dm_control/manipulator-insert_peg-v0 dm_control/pendulum-swingup-v0 dm_control/point_mass-easy-v0 dm_control/point_mass-hard-v0 dm_control/quadruped-walk-v0 dm_control/quadruped-run-v0 dm_control/quadruped-escape-v0 dm_control/quadruped-fetch-v0 dm_control/reacher-easy-v0 dm_control/reacher-hard-v0 dm_control/stacker-stack_2-v0 dm_control/stacker-stack_4-v0 dm_control/swimmer-swimmer6-v0 dm_control/swimmer-swimmer15-v0 dm_control/walker-stand-v0 dm_control/walker-walk-v0 dm_control/walker-run-v0 \
     --command "poetry run python cleanrl/ppo_continuous_action.py --exp-name ppo_continuous_action_slurm --cuda False --track" \
     --num-seeds 3 \
     --workers 9
```

I would however say that the name `ppo_continuous_action_slurm` is quite hacky... If it is not too much trouble it maybe preferable to re-run the experiments with the names `ppo_continuous_action_8M`, but it's not a big deal to me if you want to keep the current ones. We just need to add a short explanation of it in the docs.

masud99r commented:
Re-running would not be a problem.
However, I want to make sure I understand it fully. The name in the experiments needs to be `ppo_continuous_action_8M` instead of `ppo_continuous_action_slurm`. Is that right? 
In that case, for the benchmarking, it is easier for me to rename `ppo_continuous_action_slurm.py` to `ppo_continuous_action_8M.py`. This is because I need to store the experiment data in a storage-only file system. 
I will edit the command in `ppo.sh` as you mention.

vwxyzjn commented:
> Re-running would not be a problem.
> However, I want to make sure I understand it fully. The name in the experiments needs to be `ppo_continuous_action_8M` instead of `ppo_continuous_action_slurm`. Is that right?

That’s right. The `exp_name` needs to be `ppo_continuous_action_8M` for filtering.

> In that case, for the benchmarking, it is easier for me to rename `ppo_continuous_action_slurm.py` to `ppo_continuous_action_8M.py`. This is because I need to store the experiment data in a storage-only file system.

Sounds good. Thanks.

masud99r commented:
Results on all dm_control environments. The PPO and RPO run for 8M timesteps, and results are computed over 10 random seeds.

**Table:**
|                                       | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------------------------|:-------------------------------------------------------------|:----------------------------------------------|
| dm_control/acrobot-swingup-v0         | 25.12 ± 7.77                                                 | 41.62 ± 5.26                                  |
| dm_control/acrobot-swingup_sparse-v0  | 1.71 ± 0.74                                                  | 3.25 ± 0.91                                   |
| dm_control/ball_in_cup-catch-v0       | 937.71 ± 8.68                                                | 941.16 ± 9.67                                 |
| dm_control/cartpole-balance-v0        | 791.93 ± 13.67                                               | 797.17 ± 9.63                                 |
| dm_control/cartpole-balance_sparse-v0 | 990.23 ± 7.34                                                | 989.03 ± 5.20                                 |
| dm_control/cartpole-swingup-v0        | 583.60 ± 13.82                                               | 615.99 ± 11.85                                |
| dm_control/cartpole-swingup_sparse-v0 | 240.45 ± 299.08                                              | 526.18 ± 190.39                               |
| dm_control/cartpole-two_poles-v0      | 217.71 ± 5.15                                                | 219.31 ± 6.76                                 |
| dm_control/cartpole-three_poles-v0    | 159.93 ± 2.12                                                | 160.10 ± 2.36                                 |
| dm_control/cheetah-run-v0             | 473.61 ± 107.87                                              | 562.38 ± 59.67                                |
| dm_control/dog-stand-v0               | 332.40 ± 24.13                                               | 504.45 ± 135.58                               |
| dm_control/dog-walk-v0                | 124.68 ± 23.05                                               | 166.50 ± 45.85                                |
| dm_control/dog-trot-v0                | 80.08 ± 13.04                                                | 115.28 ± 30.15                                |
| dm_control/dog-run-v0                 | 67.93 ± 6.84                                                 | 104.17 ± 24.32                                |
| dm_control/dog-fetch-v0               | 28.73 ± 4.70                                                 | 44.43 ± 7.21                                  |
| dm_control/finger-spin-v0             | 628.08 ± 253.43                                              | 849.49 ± 23.43                                |
| dm_control/finger-turn_easy-v0        | 248.87 ± 80.39                                               | 447.46 ± 129.80                               |
| dm_control/finger-turn_hard-v0        | 82.28 ± 34.94                                                | 238.67 ± 134.88                               |
| dm_control/fish-upright-v0            | 541.82 ± 64.54                                               | 801.81 ± 35.79                                |
| dm_control/fish-swim-v0               | 84.81 ± 6.90                                                 | 139.70 ± 40.24                                |
| dm_control/hopper-stand-v0            | 3.11 ± 1.63                                                  | 408.24 ± 203.25                               |
| dm_control/hopper-hop-v0              | 5.84 ± 17.28                                                 | 63.69 ± 87.18                                 |
| dm_control/humanoid-stand-v0          | 21.13 ± 30.14                                                | 140.61 ± 58.36                                |
| dm_control/humanoid-walk-v0           | 8.32 ± 17.00                                                 | 77.51 ± 50.81                                 |
| dm_control/humanoid-run-v0            | 5.49 ± 9.20                                                  | 24.16 ± 19.90                                 |
| dm_control/humanoid-run_pure_state-v0 | 1.10 ± 0.13                                                  | 3.32 ± 2.50                                   |
| dm_control/humanoid_CMU-stand-v0      | 4.65 ± 0.30                                                  | 4.34 ± 0.27                                   |
| dm_control/humanoid_CMU-run-v0        | 0.86 ± 0.08                                                  | 0.85 ± 0.04                                   |
| dm_control/manipulator-bring_ball-v0  | 0.41 ± 0.22                                                  | 0.55 ± 0.37                                   |
| dm_control/manipulator-bring_peg-v0   | 0.57 ± 0.24                                                  | 2.64 ± 2.19                                   |
| dm_control/manipulator-insert_ball-v0 | 41.53 ± 15.58                                                | 39.24 ± 17.19                                 |
| dm_control/manipulator-insert_peg-v0  | 49.89 ± 14.26                                                | 53.23 ± 16.02                                 |
| dm_control/pendulum-swingup-v0        | 464.82 ± 379.09                                              | 776.21 ± 20.15                                |
| dm_control/point_mass-easy-v0         | 530.17 ± 262.14                                              | 652.18 ± 21.93                                |
| dm_control/point_mass-hard-v0         | 132.29 ± 69.71                                               | 184.29 ± 24.17                                |
| dm_control/quadruped-walk-v0          | 239.11 ± 85.20                                               | 600.71 ± 213.40                               |
| dm_control/quadruped-run-v0           | 165.70 ± 43.21                                               | 375.78 ± 119.95                               |
| dm_control/quadruped-escape-v0        | 23.82 ± 10.82                                                | 68.19 ± 30.69                                 |
| dm_control/quadruped-fetch-v0         | 183.03 ± 37.50                                               | 227.47 ± 21.67                                |
| dm_control/reacher-easy-v0            | 769.10 ± 48.35                                               | 740.93 ± 51.58                                |
| dm_control/reacher-hard-v0            | 636.89 ± 77.68                                               | 584.33 ± 43.76                                |
| dm_control/stacker-stack_2-v0         | 61.73 ± 17.34                                                | 64.32 ± 7.15                                  |
| dm_control/stacker-stack_4-v0         | 71.68 ± 29.03                                                | 53.78 ± 20.09                                 |
| dm_control/swimmer-swimmer6-v0        | 152.82 ± 30.77                                               | 166.89 ± 14.49                                |
| dm_control/swimmer-swimmer15-v0       | 148.59 ± 22.63                                               | 153.45 ± 17.12                                |
| dm_control/walker-stand-v0            | 453.73 ± 217.72                                              | 736.72 ± 145.16                               |
| dm_control/walker-walk-v0             | 308.91 ± 87.85                                               | 785.61 ± 133.29                               |
| dm_control/walker-run-v0              | 123.98 ± 91.34                                               | 384.08 ± 123.89                               |

**Learning curves:**
![dm_control_all_ppo_rpo_8M](https://user-images.githubusercontent.com/15100273/209252957-96bcee97-18ac-47a4-86ed-abffcbba3952.png)

vwxyzjn commented:
The results look exceptionally clean 👍. After all the reported results, I would feel comfortable marking the existing `ppo_continuous_action.py` as "deprecated" or "not recommended" in favor of `rpo_continuous_action.py` because of the simplicity of the approach and the superiority of the empirical results. 

Please update the docs with these results. Please also remove the `ppo_continuous_action_slurm.py` results from the project, as they are no longer needed.

Thank you, and have a great holiday.

Howuhh commented:
Seeing how much this improves the scores, I'm now really curious what happens if we change from Gaussian to the Beta distribution...

masud99r commented:
@Howuhh Trying Beta distribution would be an interesting idea. I observed some performance effects when trying with Laplace and Gumbel (https://arxiv.org/pdf/2212.07536.pdf Figure 7 & 8). 

An aspect of the RPO method is maintaining randomness through Uniform distribution. It would be interesting to see if clip log std can achieve a similar effect.

masud99r commented:
Hi @vwxyzjn, I have added a doc file named [`rpo.md`](https://github.com/masud99r/cleanrl/blob/master/docs/rl-algorithms/rpo.md). Please review it when you have some time.

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,129 @@
+| dm_control/cartpole-three_poles-v0    | 159.93 ± 2.12                                                | 160.10 ± 2.36                                 |
+| dm_control/cheetah-run-v0             | 473.61 ± 107.87                                              | 562.38 ± 59.67                                |
+| dm_control/dog-stand-v0               | 332.40 ± 24.13                                               | 504.45 ± 135.58                               |
+| dm_control/dog-walk-v0                | 124.68 ± 23.05                                               | 166.50 ± 45.85                                |
+| dm_control/dog-trot-v0                | 80.08 ± 13.04                                                | 115.28 ± 30.15                                |
+| dm_control/dog-run-v0                 | 67.93 ± 6.84                                                 | 104.17 ± 24.32                                |
+| dm_control/dog-fetch-v0               | 28.73 ± 4.70                                                 | 44.43 ± 7.21                                  |
+| dm_control/finger-spin-v0             | 628.08 ± 253.43                                              | 849.49 ± 23.43                                |
+| dm_control/finger-turn_easy-v0        | 248.87 ± 80.39                                               | 447.46 ± 129.80                               |
+| dm_control/finger-turn_hard-v0        | 82.28 ± 34.94                                                | 238.67 ± 134.88                               |
+| dm_control/fish-upright-v0            | 541.82 ± 64.54                                               | 801.81 ± 35.79                                |
+| dm_control/fish-swim-v0               | 84.81 ± 6.90                                                 | 139.70 ± 40.24                                |
+| dm_control/hopper-stand-v0            | 3.11 ± 1.63                                                  | 408.24 ± 203.25                               |
+| dm_control/hopper-hop-v0              | 5.84 ± 17.28                                                 | 63.69 ± 87.18                                 |
+| dm_control/humanoid-stand-v0          | 21.13 ± 30.14                                                | 140.61 ± 58.36                                |
+| dm_control/humanoid-walk-v0           | 8.32 ± 17.00                                                 | 77.51 ± 50.81                                 |
+| dm_control/humanoid-run-v0            | 5.49 ± 9.20                                                  | 24.16 ± 19.90                                 |
+| dm_control/humanoid-run_pure_state-v0 | 1.10 ± 0.13                                                  | 3.32 ± 2.50                                   |
+| dm_control/humanoid_CMU-stand-v0      | 4.65 ± 0.30                                                  | 4.34 ± 0.27                                   |
+| dm_control/humanoid_CMU-run-v0        | 0.86 ± 0.08                                                  | 0.85 ± 0.04                                   |
+| dm_control/manipulator-bring_ball-v0  | 0.41 ± 0.22                                                  | 0.55 ± 0.37                                   |
+| dm_control/manipulator-bring_peg-v0   | 0.57 ± 0.24                                                  | 2.64 ± 2.19                                   |
+| dm_control/manipulator-insert_ball-v0 | 41.53 ± 15.58                                                | 39.24 ± 17.19                                 |
+| dm_control/manipulator-insert_peg-v0  | 49.89 ± 14.26                                                | 53.23 ± 16.02                                 |
+| dm_control/pendulum-swingup-v0        | 464.82 ± 379.09                                              | 776.21 ± 20.15                                |
+| dm_control/point_mass-easy-v0         | 530.17 ± 262.14                                              | 652.18 ± 21.93                                |
+| dm_control/point_mass-hard-v0         | 132.29 ± 69.71                                               | 184.29 ± 24.17                                |
+| dm_control/quadruped-walk-v0          | 239.11 ± 85.20                                               | 600.71 ± 213.40                               |
+| dm_control/quadruped-run-v0           | 165.70 ± 43.21                                               | 375.78 ± 119.95                               |
+| dm_control/quadruped-escape-v0        | 23.82 ± 10.82                                                | 68.19 ± 30.69                                 |
+| dm_control/quadruped-fetch-v0         | 183.03 ± 37.50                                               | 227.47 ± 21.67                                |
+| dm_control/reacher-easy-v0            | 769.10 ± 48.35                                               | 740.93 ± 51.58                                |
+| dm_control/reacher-hard-v0            | 636.89 ± 77.68                                               | 584.33 ± 43.76                                |
+| dm_control/stacker-stack_2-v0         | 61.73 ± 17.34                                                | 64.32 ± 7.15                                  |
+| dm_control/stacker-stack_4-v0         | 71.68 ± 29.03                                                | 53.78 ± 20.09                                 |
+| dm_control/swimmer-swimmer6-v0        | 152.82 ± 30.77                                               | 166.89 ± 14.49                                |
+| dm_control/swimmer-swimmer15-v0       | 148.59 ± 22.63                                               | 153.45 ± 17.12                                |
+| dm_control/walker-stand-v0            | 453.73 ± 217.72                                              | 736.72 ± 145.16                               |
+| dm_control/walker-walk-v0             | 308.91 ± 87.85                                               | 785.61 ± 133.29                               |
+| dm_control/walker-run-v0              | 123.98 ± 91.34                                               | 384.08 ± 123.89                               |
```
A couple of comments:

1. The table formatting is broken at https://cleanrl-git-fork-masud99r-master-vwxyzjn.vercel.app/rl-algorithms/rpo/#experiment-results?
2. When displaying the gymnasium results, we usually display them in tabs such as https://docs.cleanrl.dev/rl-algorithms/ppo/#experiment-results_2. See the source code at https://github.com/vwxyzjn/cleanrl/blob/master/docs/rl-algorithms/ppo.md#experiment-results-2.
3. Do you have the mujoco results on `HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4 Pusher-v4`?

masud99r replied:
Thanks for the thorough feedback. I will make sure to update the docs accordingly (1 & 2).

3. As for Mujoco results, I still need to run them and see how RPO performs. So I will post the results here and update the docs. These results will help in deciding the depreciation decision of  `ppo_continuous_action.py`. 

Typically, RPO shows improved results compared to PPO in difficult environments. It seems PPO is performing quite well already on many Mujoco environments. I will start running the experiments to see how the performance stands.

masud99r replied:
Where can I find a list of all the mujoco v2 and v4 environments? @vwxyzjn

vwxyzjn replied:
That makes sense. I think in the mujoco environments, RPO should not make the performance worse. Let's run the v2 environments as well just to be sure:

```
poetry install --with mujoco_py,mujoco
poetry run python -c "import mujoco_py"
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids HalfCheetah-v2 Walker2d-v2 Hopper-v2 InvertedPendulum-v2 Humanoid-v2 Pusher-v2 \
    --command "poetry run python cleanrl/ppo_continuous_action.py --cuda False --track --capture-video" \
    --num-seeds 3 \
    --workers 1
poetry install --with mujoco
OMP_NUM_THREADS=1 xvfb-run -a python -m cleanrl_utils.benchmark \
    --env-ids HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4 Pusher-v4 \
    --command "poetry run python cleanrl/gymnasium_support/ppo_continuous_action.py --cuda False --track" \
    --num-seeds 3 \
    --workers 1
```

The full mujoco environments are at https://gymnasium.farama.org/environments/mujoco/. Feel free to run all of them :), which would be


Ant-v2 HalfCheetah-v2 Hopper-v2 HumanoidStandup-v2 Humanoid-v2 InvertedDoublePendulum-v2 InvertedPendulum-v2 Reacher-v2 Swimmer-v2 Pusher-v2 Walker2d-v2  

and 

Ant-v4 HalfCheetah-v4 Hopper-v4 HumanoidStandup-v4 Humanoid-v4 InvertedDoublePendulum-v4 InvertedPendulum-v4 Reacher-v4 Swimmer-v4 Pusher-v4 Walker2d-v4

masud99r replied:
Talking about environments, would you like to benchmark results on PyBullet? Do you have a preferred list of environments? Here is a list I find: https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/gym/pybullet_envs/__init__.py

vwxyzjn replied:
I wouldn't say no to the pybullet experiments, but previously they were omitted because pybullet seemed to be no longer maintained. It's also the case that it needs an adapter for gymnasium 

https://github.com/bulletphysics/bullet3/issues/4369, https://github.com/bulletphysics/bullet3/pull/4335. However, it is possible to fix this issue with a compatibility wrapper (https://gymnasium.farama.org/api/wrappers/misc_wrappers/#gymnasium.wrappers.EnvCompatibility).  

If you would like to add the pybullet results, feel free to use your preferred list of experiments and add another tab in the experiment section to report the results. 
<img width="783" alt="image" src="https://user-images.githubusercontent.com/5555347/209744917-6cf9d08c-238c-4edf-b91c-4c415d640b85.png">

masud99r replied:
Results are in for mujoco (v4, and v2). Some environments require tuning of RPO `alpha` parameter as described below:

**Mujoco v4:**

Table:

|                     | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------|:-------------------------------------------------------------|:----------------------------------------------|
| HumanoidStandup-v4  | 107841.91 ± 15232.17                                         | 150129.37 ± 6447.36                           |
| Humanoid-v4         | 590.33 ± 29.38                                               | 763.79 ± 137.91                               |
| InvertedPendulum-v4 | 883.61 ± 25.01                                               | 875.56 ± 39.68                                |
| Walker2d-v4         | 2872.50 ± 683.82                                             | 3707.25 ± 346.22                              |

Learning curves:

![mujoco_v4_part1](https://user-images.githubusercontent.com/15100273/209876730-da7cde5f-1260-4009-96dd-8bbb563c40a8.png)

**Mujoco v2:**

Table:

|                     | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------|:-------------------------------------------------------------|:----------------------------------------------|
| HumanoidStandup-v2  | 108982.78 ± 17996.44                                         | 156755.15 ± 10672.52                          |
| Humanoid-v2         | 586.11 ± 44.21                                               | 723.01 ± 103.07                               |
| InvertedPendulum-v2 | 861.44 ± 27.30                                               | 857.88 ± 30.67                                |
| Walker2d-v2         | 3219.50 ± 948.84                                             | 3934.37 ± 428.82                              |

Learning curves:

![mujoco_v2_part1](https://user-images.githubusercontent.com/15100273/209876825-33bfcedf-5afd-4878-a117-5833d2ce965e.png)


The following environments require tuning of `alpha` (Algorithm 1, line 13, paper: https://arxiv.org/pdf/2212.07536.pdf). As described in the paper, this variable should be tuned for environments tested. A larger value means more randomness, whereas a smaller value indicates less randomness. Some mujoco environments require a smaller `alpha=0.01` value to achieve a reasonable performance compared to `alpha=0.5` for the rest of the environments. This version (`alpha=0.01`) of runs is indicated as `rpo_continuous_action_alpha_0_01` in the table and learning curves.

**Mujoco v4:**

Table:

|                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action_alpha_0_01 ({'tag': ['pr-331']})   |
|:--------------------------|:-------------------------------------------------------------|:---------------------------------------------------------|
| Ant-v4                    | 1777.42 ± 844.67                                             | 2704.84 ± 665.91                                         |
| HalfCheetah-v4            | 2654.93 ± 1098.90                                            | 2729.61 ± 1311.41                                        |
| Hopper-v4                 | 2784.27 ± 280.23                                             | 2438.67 ± 492.42                                         |
| InvertedDoublePendulum-v4 | 5586.45 ± 213.17                                             | 5406.54 ± 269.84                                         |
| Reacher-v4                | -4.66 ± 0.84                                                 | -3.97 ± 0.17                                             |
| Swimmer-v4                | 124.45 ± 22.31                                               | 130.25 ± 12.48                                           |
| Pusher-v4                 | -30.54 ± 6.55                                                | -31.54 ± 9.80                                            |

Learning curves:

![mujoco_v4_part2](https://user-images.githubusercontent.com/15100273/209877045-c2d06cf7-c693-41aa-a763-3b4bc8592325.png)


**Mujoco v2:**

Table:

|                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action_alpha_0_01 ({'tag': ['pr-331']})   |
|:--------------------------|:-------------------------------------------------------------|:---------------------------------------------------------|
| Ant-v2                    | 2435.03 ± 955.57                                             | 3142.98 ± 787.87                                         |
| HalfCheetah-v2            | 2714.14 ± 1260.34                                            | 2716.46 ± 1241.38                                        |
| Hopper-v2                 | 2411.13 ± 657.94                                             | 2286.05 ± 590.10                                         |
| InvertedDoublePendulum-v2 | 5727.86 ± 367.86                                             | 5811.94 ± 281.47                                         |
| Reacher-v2                | -4.57 ± 0.48                                                 | -4.17 ± 0.22                                             |
| Swimmer-v2                | 132.43 ± 10.01                                               | 141.49 ± 8.47                                            |
| Pusher-v2                 | -33.60 ± 8.54                                                | -26.08 ± 2.32                                            |

Learning curves:

![mujoco_v2_part2](https://user-images.githubusercontent.com/15100273/209877085-0f499aad-3117-4bf7-8ee2-2339f13c7c0b.png)

masud99r replied:
As for the PyBullet, let's skip the experiments for now.

vwxyzjn replied:
These results look very nice! Did you run the utility with `--scan-history` for the v2 charts? The charts looked way too smooth compared to the v4 ones.

masud99r replied:
I keep all the arguments the same. However, I had to uncomment ` n_samples=500,`, otherwise, I was getting an error message. Do you think that might cause the smoothness? 
Here is the error message:
```    uniques, codes = table.factorize(
  File "pandas/_libs/hashtable_class_helper.pxi", line 5943, in pandas._libs.hashtable.PyObjectHashTable.factorize
  File "pandas/_libs/hashtable_class_helper.pxi", line 5857, in pandas._libs.hashtable.PyObjectHashTable._unique
TypeError: unhashable type: 'dict'```

vwxyzjn replied:
I just ran the following and it gave me 
```
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6' 'rpo_continuous_action_alpha_0_01?tag=pr-331' \
    --env-ids HalfCheetah-v2 Walker2d-v2 Hopper-v2 InvertedPendulum-v2 Humanoid-v2 Pusher-v2 \
    --check-empty-runs False \
    --ncols 4 \
    --ncols-legend 3 \
    --output-filename compare \
    --scan-history
```

<img width="1467" alt="image" src="https://user-images.githubusercontent.com/5555347/210016522-ff94a6cf-8f07-4da3-be55-1d90b6e2c901.png">

Maybe try upgrading by doing `pip install openrlbenchmark --upgrade`

vwxyzjn commented on benchmark/rpo.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+export WANDB_ENTITY=openrlbenchmark
+
```
Remove these

vwxyzjn commented on benchmark/rpo.sh:
```diff
@@ -0,0 +1,47 @@
+#     --num-seeds 1 \
+#     --workers 1
+
+xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
+    --env-ids $1\
+    --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
+    --num-seeds 1 \
+    --start-seed $2 \
+    --workers 1
+
+
+
+
+
+
+# poetry install
+# OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
+#     --env-ids Pendulum-v1 BipedalWalker-v3\
+#     --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
+#     --num-seeds 10 \
+#     --workers 20
+
+# OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
+#     --env-ids BipedalWalker-v3\
+#     --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
+#     --num-seeds 3 \
+#     --workers 1
+
+# dm_control
+# OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
+#     --env-ids dm_control/hopper-hop-v0\
+#     --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
+#     --num-seeds 3 \
+#     --workers 1
+
+# OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
+#     --env-ids dm_control/acrobot-swingup-v0 dm_control/acrobot-swingup_sparse-v0 dm_control/ball_in_cup-catch-v0 dm_control/cartpole-balance-v0 dm_control/cartpole-balance_sparse-v0 dm_control/cartpole-swingup-v0 dm_control/cartpole-swingup_sparse-v0 dm_control/cartpole-two_poles-v0 dm_control/cartpole-three_poles-v0 dm_control/cheetah-run-v0 dm_control/dog-stand-v0 dm_control/dog-walk-v0 dm_co ...
+#     --command "poetry run python cleanrl/gymnasium_support/ppo_continuous_action.py --cuda False --track" \
+#     --num-seeds 3 \
+#     --workers 9
```
Please clean up here. You should probably have

```
# poetry install command to be filled here
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids Pendulum-v1 BipedalWalker-v3 \
    --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
    --num-seeds 1 \
    --workers 1

poetry install --with mujoco
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4 Pusher-v4 \
    --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track --capture-video" \
    --num-seeds 10 \
    --workers 1

poetry install --with mujoco,dm_control
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids dm_control/acrobot-swingup-v0 dm_control/acrobot-swingup_sparse-v0 dm_control/ball_in_cup-catch-v0 dm_control/cartpole-balance-v0 dm_control/cartpole-balance_sparse-v0 dm_control/cartpole-swingup-v0 dm_control/cartpole-swingup_sparse-v0 dm_control/cartpole-two_poles-v0 dm_control/cartpole-three_poles-v0 dm_control/cheetah-run-v0 dm_control/dog-stand-v0 dm_control/dog-walk-v0 dm_control/dog-trot-v0 dm_control/dog-run-v0 dm_control/dog-fetch-v0 dm_control/finger-spin-v0 dm_control/finger-turn_easy-v0 dm_control/finger-turn_hard-v0 dm_control/fish-upright-v0 dm_control/fish-swim-v0 dm_control/hopper-stand-v0 dm_control/hopper-hop-v0 dm_control/humanoid-stand-v0 dm_control/humanoid-walk-v0 dm_control/humanoid-run-v0 dm_control/humanoid-run_pure_state-v0 dm_control/humanoid_CMU-stand-v0 dm_control/humanoid_CMU-run-v0 dm_control/lqr-lqr_2_1-v0 dm_control/lqr-lqr_6_2-v0 dm_control/manipulator-bring_ball-v0 dm_control/manipulator-bring_peg-v0 dm_control/manipulator-insert_ball-v0 dm_control/manipulator-insert_peg-v0 dm_control/pendulum-swingup-v0 dm_control/point_mass-easy-v0 dm_control/point_mass-hard-v0 dm_control/quadruped-walk-v0 dm_control/quadruped-run-v0 dm_control/quadruped-escape-v0 dm_control/quadruped-fetch-v0 dm_control/reacher-easy-v0 dm_control/reacher-hard-v0 dm_control/stacker-stack_2-v0 dm_control/stacker-stack_4-v0 dm_control/swimmer-swimmer6-v0 dm_control/swimmer-swimmer15-v0 dm_control/walker-stand-v0 dm_control/walker-walk-v0 dm_control/walker-run-v0 \
    --command "poetry run python cleanrl/rpo_continuous_action.py --cuda False --track" \
    --num-seeds 10 \
    --workers 1
```

and have the following in the `ppo.sg`

```bash
poetry install --with mujoco,dm_control
OMP_NUM_THREADS=1 xvfb-run -a poetry run python -m cleanrl_utils.benchmark \
    --env-ids dm_control/acrobot-swingup-v0 dm_control/acrobot-swingup_sparse-v0 dm_control/ball_in_cup-catch-v0 dm_control/cartpole-balance-v0 dm_control/cartpole-balance_sparse-v0 dm_control/cartpole-swingup-v0 dm_control/cartpole-swingup_sparse-v0 dm_control/cartpole-two_poles-v0 dm_control/cartpole-three_poles-v0 dm_control/cheetah-run-v0 dm_control/dog-stand-v0 dm_control/dog-walk-v0 dm_control/dog-trot-v0 dm_control/dog-run-v0 dm_control/dog-fetch-v0 dm_control/finger-spin-v0 dm_control/finger-turn_easy-v0 dm_control/finger-turn_hard-v0 dm_control/fish-upright-v0 dm_control/fish-swim-v0 dm_control/hopper-stand-v0 dm_control/hopper-hop-v0 dm_control/humanoid-stand-v0 dm_control/humanoid-walk-v0 dm_control/humanoid-run-v0 dm_control/humanoid-run_pure_state-v0 dm_control/humanoid_CMU-stand-v0 dm_control/humanoid_CMU-run-v0 dm_control/lqr-lqr_2_1-v0 dm_control/lqr-lqr_6_2-v0 dm_control/manipulator-bring_ball-v0 dm_control/manipulator-bring_peg-v0 dm_control/manipulator-insert_ball-v0 dm_control/manipulator-insert_peg-v0 dm_control/pendulum-swingup-v0 dm_control/point_mass-easy-v0 dm_control/point_mass-hard-v0 dm_control/quadruped-walk-v0 dm_control/quadruped-run-v0 dm_control/quadruped-escape-v0 dm_control/quadruped-fetch-v0 dm_control/reacher-easy-v0 dm_control/reacher-hard-v0 dm_control/stacker-stack_2-v0 dm_control/stacker-stack_4-v0 dm_control/swimmer-swimmer6-v0 dm_control/swimmer-swimmer15-v0 dm_control/walker-stand-v0 dm_control/walker-walk-v0 dm_control/walker-run-v0 \
    --command "poetry run python cleanrl/ppo_continuous_action.py --exp-name ppo_continuous_action_8M  --cuda False --track" \
    --num-seeds 10 \
    --workers 1
```

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,129 @@
+See [related docs](/rl-algorithms/ppo/#explanation-of-the-logged-metrics) for `ppo.py`.
+
+### Implementation details
+Similar to PPO [ppo_continuous_action.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py) with a change in `get_action_and_value` method in `Agent` class.
```
You can write it as 

```md
[rpo_continuous_action.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/rpo_continuous_action.py) has the same implementation details as `ppo_continuous_action.py` (see related [docs](/rl-algorithms/ppo/#ppo_continuous_actionpy)) but with a few lines of code differences.
```

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,129 @@
+### Implementation details
+Similar to PPO [ppo_continuous_action.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py) with a change in `get_action_and_value` method in `Agent` class.
+
+[TODO: Add code snippets, and comparison with PPO]
```
Please fill this up. You can leverage the highlighted lines in the code blocks to illustrate the code differences https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#highlighting-specific-lines. 

We did something similar for DDPG.
<img width="811" alt="image" src="https://user-images.githubusercontent.com/5555347/209712882-8e25b5d9-2499-4e80-9c27-72b963ca8f16.png">

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,129 @@
+
+[TODO: Add code snippets, and comparison with PPO]
+
+### Experiment results
+
+**DeepMind Control**
+
+Results on all dm_control environments. The PPO and RPO run for 8M timesteps, and results are computed over 10 random seeds.
```
Please add a GitHub snippet embed pointing to `rpo.sh` to show the benchmark commands to help with reproducibility, like the following. Feel free to take a look at other docs as an example.

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/5555347/209712999-6636b19b-c80e-4d6f-81ae-740e3104ce45.png">

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,129 @@
+Similar to PPO [ppo_continuous_action.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py) with a change in `get_action_and_value` method in `Agent` class.
+
+[TODO: Add code snippets, and comparison with PPO]
+
```
In the implementation details section, you should also add a note on the difference on hyperparameters. We used a table for `ppo_continuous_action_isaacgym.py`, but maybe in this case you just add a note since there is only one hyperparameter difference (--total-timesteps=8000000)

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/5555347/209713230-11658d88-fed1-49d7-a8c1-c9555e7d3546.png">

vwxyzjn reviewed (changes requested):
Excellent work! @masud99r thanks for helping with the documentation. Some comments are below, in addition to the inline reviews.

## poetry and `BipedalWalker-v3`

If we want to include the `BipedalWalker-v3` experiments, we should also update the dependencies to include box2d-py for reproducibility purposes, but that seems to cause an issue for poetry (see https://github.com/Farama-Foundation/Gymnasium/pull/86#issuecomment-1294321247). What we can do instead is to specify the installation in the `Usage` section of the docs. Something like

```
# mujoco v4 environments
poetry install --with mujoco
python cleanrl/rpo_continuous_action.py --help
python cleanrl/rpo_continuous_action.py --env-id Hopper-v2
# dm_control v4 environments
poetry install --with mujoco,dm_control
python cleanrl/rpo_continuous_action.py --env-id dm_control/cartpole-balance-v0
# BipedalWalker-v3 experiment (hack)
poetry install
poetry run pip install box2d-py==some-version
python cleanrl/rpo_continuous_action.py --env-id BipedalWalker-v3
```

## A note on deprecating `ppo_continuous_action.py`

We should add a note to deprecate `ppo_continuous_action.py`, probably by using the warning blocks like the following. Note that to deprecate `ppo_continuous_action.py` we should also need the mujoco experiments to show `rpo_continuous_action.py`'s better empirical results, and in the notes we can say "we are deprecating `ppo_continuous_action.py` because `rpo_continuous_action.py` is a **strictly better** algorithm"

<img width="783" alt="image" src="https://user-images.githubusercontent.com/5555347/209714475-b9ce1817-8beb-482c-8856-2c0cc32e188a.png">


Todo items:
- [ ] poetry stuff
- [ ] add a note to deprecate `ppo_continuous_action.py` in favor of `rpo_continuous_action.py`

masud99r commented:
Resolved. Thank you for the suggestion.

> ## poetry and `BipedalWalker-v3`
> If we want to include the `BipedalWalker-v3` experiments, we should also update the dependencies to include box2d-py for reproducibility purposes, but that seems to cause an issue for poetry (see [Farama-Foundation/Gymnasium#86 (comment)](https://github.com/Farama-Foundation/Gymnasium/pull/86#issuecomment-1294321247)). What we can do instead is to specify the installation in the `Usage` section of the docs. Something like
> 
> ```
> # mujoco v4 environments
> poetry install --with mujoco
> python cleanrl/rpo_continuous_action.py --help
> python cleanrl/rpo_continuous_action.py --env-id Hopper-v2
> # dm_control v4 environments
> poetry install --with mujoco,dm_control
> python cleanrl/rpo_continuous_action.py --env-id dm_control/cartpole-balance-v0
> # BipedalWalker-v3 experiment (hack)
> poetry install
> poetry run pip install box2d-py==some-version
> python cleanrl/rpo_continuous_action.py --env-id BipedalWalker-v3
> ```

vwxyzjn commented on cleanrl/rpo_continuous_action.py:
```diff
@@ -0,0 +1,338 @@
+            action = probs.sample()
+        else: # new to RPO
+            # sample again to add stochasticity, for the policy update
+            z = torch.FloatTensor(action_mean.shape).uniform_(-0.5, 0.5)
```
You should make this `alpha` to be a hyperparameter. Something like

```python
    parser.add_argument("--target-kl", type=float, default=None,
        help="the target KL divergence threshold")
    parser.add_argument("--rpo-alpha", type=float, default=0.5,
        help="the alpha parameter for RPO")


class Agent(nn.Module):
    def __init__(self, envs):
        super().__init__()
        self.critic = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), 64)),
            nn.Tanh(),
            layer_init(nn.Linear(64, 64)),
            nn.Tanh(),
            layer_init(nn.Linear(64, 1), std=1.0),
        )
        self.actor_mean = nn.Sequential(
            layer_init(nn.Linear(np.array(envs.single_observation_space.shape).prod(), 64)),
            nn.Tanh(),
            layer_init(nn.Linear(64, 64)),
            nn.Tanh(),
            layer_init(nn.Linear(64, np.prod(envs.single_action_space.shape)), std=0.01),
        )
        self.actor_logstd = nn.Parameter(torch.zeros(1, np.prod(envs.single_action_space.shape)))

    def get_value(self, x):
        return self.critic(x)

    def get_action_and_value(self, x, action=None):
        action_mean = self.actor_mean(x)
        action_logstd = self.actor_logstd.expand_as(action_mean)
        action_std = torch.exp(action_logstd)
        probs = Normal(action_mean, action_std)
        if action is None:
            action = probs.sample()
        else: # new to RPO
            # sample again to add stochasticity, for the policy update
            z = torch.FloatTensor(action_mean.shape).uniform_(-self.rpo_alpha, self.rpo_alpha)
            action_mean = action_mean + z
            probs = Normal(action_mean, action_std)

        return action, probs.log_prob(action).sum(1), probs.entropy().sum(1), self.critic(x)
```

You should also update the `rpo.sh` correspondingly.

masud99r replied:
Yes, I agree, great suggestion.
Is `self.rpo_alpha` available from `get_action_and_value` method?

vwxyzjn replied:
Sorry I meant it should be part of the constructor 

```
class Agent(nn.Module):
    def __init__(self, envs, rpo_alpha):
        super().__init__()
        ...
        self.rpo_alpha = rpo_alpha

agent = Agent(envs, args.rpo_alpha)
```

masud99r replied:
Got it. It would have been better if we could do it without the constructor; then, the code difference with PPO would be less. However, I will follow the constructor approach for now.

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,190 @@
+        Learning curves:
+        ![](../rpo/mujoco_v2_part1.png)
+
+        The following environments require tuning of `alpha` (Algorithm 1, line 13, paper: https://arxiv.org/pdf/2212.07536.pdf). As described in the paper, this variable should be tuned for environments tested. A larger value means more randomness, whereas a smaller value indicates less randomness. Some mujoco environments require a smaller `alpha=0.01` value to achieve a reasonable performance  ...
+
+
+        |                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action_alpha_0_01 ({'tag': ['pr-331']})   |
+        |:--------------------------|:-------------------------------------------------------------|:---------------------------------------------------------|
+        | Ant-v2                    | 2435.03 ± 955.57                                             | 3142.98 ± 787.87                                         |
+        | HalfCheetah-v2            | 2714.14 ± 1260.34                                            | 2716.46 ± 1241.38                                        |
+        | Hopper-v2                 | 2411.13 ± 657.94                                             | 2286.05 ± 590.10                                         |
+        | InvertedDoublePendulum-v2 | 5727.86 ± 367.86                                             | 5811.94 ± 281.47                                         |
+        | Reacher-v2                | -4.57 ± 0.48                                                 | -4.17 ± 0.22                                             |
+        | Swimmer-v2                | 132.43 ± 10.01                                               | 141.49 ± 8.47                                            |
+        | Pusher-v2                 | -33.60 ± 8.54                                                | -26.08 ± 2.32                                            |
```
This is interesting. Do you have the results for `alpha=0.5` for these environments? Do they make the performance a lot worse?

masud99r replied:
Yes, the performance is generally worse for `alpha=0.5` for those environments. They are very sensitive to random perturbation on the mean. Here are the results for Mujoco v4 (v2 results are similar).

|                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
|:--------------------------|:-------------------------------------------------------------|:----------------------------------------------|
| Ant-v4                    | 1808.30 ± 840.75                                             | -6.33 ± 2.75                                  |
| HalfCheetah-v4            | 2621.80 ± 1065.70                                            | 2149.96 ± 787.94                              |
| Hopper-v4                 | 2718.36 ± 306.63                                             | 1561.23 ± 202.97                              |
| InvertedDoublePendulum-v4 | 5703.07 ± 339.56                                             | 301.79 ± 15.95                                |
| Reacher-v4                | -4.65 ± 0.71                                                 | -66.57 ± 0.65                                 |
| Swimmer-v4                | 124.63 ± 21.98                                               | 117.50 ± 9.64                                 |
| Pusher-v4                 | -30.35 ± 6.34                                                | -277.30 ± 26.16                               |

Learning curves:
![mujoco_v4_part2_0_05](https://user-images.githubusercontent.com/15100273/209893934-34dd321a-ee24-4afc-ba42-4673ddf22286.png)

I have results for `alpha=0.05, 0.1` in the https://wandb.ai/openrlbenchmark/cleanrl for all these environments. Typically, results get better as we lower the value of `alpha`, and perform better (even better than PPO in some environments).

vwxyzjn replied:
Thanks for sharing these results. Some thoughts:

1. These results are fine as long as the settings to reproduce them are listed in `rpo.sh` and in the docs. 
2. You should also have a section in the docs that list out these "failed" experiments (probably via a [collapsable block](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks)), that way, the readers will be aware of this issue.
3. What does the performance look like for dm_control environments when `alpha=0.05`? It would be most desirable to find a single `alpha` that works strictly equal to or better than the default PPO in 95% of the environments tested.

masud99r replied:
1 & 2: Sure, I will add them to the docs.

3. I do not have dm_control results with `alpha=0.05`. I have done an ablation study on a subset of them, and `alpha=0.5` seems to do better across these environments (Figure 17, https://arxiv.org/pdf/2212.07536.pdf). It seems to me that the choice of `alpha` depends on the environment's design (e.g., action scaling); thus, the optimal value might vary across environments (most likely across benchmarks).

masud99r replied:
As it stands now: `alpha=0.5` is strictly equal to or better than the default PPO in 89% of environments tested (all 48/48 dm_control, 2/2 Gym, 7/11 mujoco_v4, 7/11 mujoco_v2. Total 64 out of 72 environments tested).

vwxyzjn replied:
> As it stands now: alpha=0.5 is strictly equal to or better than the default PPO in 89% of environments tested (all 48/48 dm_control, 2/2 Gym, 7/11 mujoco_v4, 7/11 mujoco_v2. Total 64 out of 72 environments tested).

This makes sense. Let's put this notice (RPO with alpha=0.5 is strictly equal to or better than the default PPO in 89% of environments ) in the PPO deprecation notice.

masud99r replied:
Thoughts:
I do not see much difference between mujoco v4 and v2 in performance comparison. This is because they are the same environments. Is there any fundamental design difference between those environments? If not, then can we skip mujoco v2 environments for the comparison? It will make: RPO `alpha=0.5` strictly equal to or better than the default PPO in 93% of environments tested (all 48/48 dm_control, 2/2 Gym, 7/11 mujoco_v4. Total 57 out of 61 environments tested). Please let me know what you think. @vwxyzjn

vwxyzjn replied:
That's a good point. Then let's use this updated stats, but we probably should keep the v2 environment results for compatibility purposes :)

masud99r replied:
Yes, sure.

vwxyzjn reviewed (commented):
Left a few more comments.

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,228 @@
+        | Walker2d-v4         | 2872.50 ± 683.82                                             | 3707.25 ± 346.22                              |
+
+        Learning curves:
+        ![](../rpo/mujoco_v4_part1.png)
```
At the end of the learning curves, you should also add the tracked experiments. When invoking the `openrlbenchmark` command, you can add a `--report` to generate a report. Then you can clone this report to the [`openrlbenchmark/openrlbenchmark` project](https://wandb.ai/openrlbenchmark/openrlbenchmark/reportlist). Finally you can embed the report as an iframe in the docs, as done in the docs. The report allows the readers to track where the experiments come from and they can query additional info they are interested (e.g., hardware usage)

masud99r replied:
This is such an awesome tool you develop. So convenient to generate results for tables, curves, and reports from wandb. Awesome!!!

I will add them accordingly.

masud99r commented:
Help needed:
The docs work as expected, except when I add `iframe` to incorporate tracked experiments. When I add tracked experiments `iframe` the docs show the first iframe tracked results and break from there. Do you have any idea what I am missing here? @vwxyzjn 

RPO Doc: https://github.com/masud99r/cleanrl/blob/master/docs/rl-algorithms/rpo.md

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,344 @@
+
+        <iframe loading="lazy" src="https://wandb.ai/openrlbenchmark/cleanrl/reports/RPO-on-Mujoco_v2-Part-2--VmlldzozMjU4MzI1" style="width:100%; height:500px" title="RPO-mujoco-v2-part1"></iframe>
+
+        Results with `rpo_alpha=0.5` (not tuned) on the tuned environments:
+
+        |                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
+        |:--------------------------|:-------------------------------------------------------------|:----------------------------------------------|
+        | Ant-v2                    | 2495.65 ± 991.65                                             | -7.81 ± 3.57                                  |
+        | HalfCheetah-v2            | 2722.03 ± 1231.28                                            | 2605.06 ± 1183.30                             |
+        | Hopper-v2                 | 2356.83 ± 650.91                                             | 1609.79 ± 164.16                              |
+        | InvertedDoublePendulum-v2 | 5675.31 ± 244.34                                             | 274.78 ± 16.40                                |
+        | Reacher-v2                | -4.67 ± 0.48                                                 | -66.55 ± 0.20                                 |
+        | Swimmer-v2                | 131.53 ± 9.94                                                | 114.34 ± 3.95                                 |
+        | Pusher-v2                 | -33.46 ± 8.41                                                | -275.09 ± 15.65                               |
+
+        Learning curves:
+        ![](../rpo/mujoco_v2_part2_0_5.png)
+
+        <!-- Tracked experiments: <iframe loading="lazy" src="https://wandb.ai/openrlbenchmark/cleanrl/reports/RPO-alpha-0-5-on-Mujoco_v2-Part-2--VmlldzozMjU4MzQ0" style="border:none;height:1024px;width:100%"> -->
```
Is this duplicate info since we have the "Failure case of `rpo_alpha=0.5`:" down below?

vwxyzjn replied:
Oh nvm I see what you did there

vwxyzjn commented on docs/rl-algorithms/rpo.md:
```diff
@@ -0,0 +1,344 @@
+        <iframe loading="lazy" src="https://wandb.ai/openrlbenchmark/cleanrl/reports/RPO-on-Mujoco_v4-Part-2--VmlldzozMjU3OTM4" style="width:100%; height:500px" title="RPO-mujoco-v4-part2"></iframe>
+        
+
+        Results with `rpo_alpha=0.5` (not tuned) on the tuned environments:
+
+        |                           | ppo_continuous_action_8M ({'tag': ['v1.0.0-13-gcbd83f6']})   | rpo_continuous_action ({'tag': ['pr-331']})   |
+        |:--------------------------|:-------------------------------------------------------------|:----------------------------------------------|
+        | Ant-v4                    | 1774.42 ± 819.08                                             | -7.99 ± 2.47                                  |
+        | HalfCheetah-v4            | 2667.34 ± 1109.99                                            | 2163.57 ± 790.16                              |
+        | Hopper-v4                 | 2761.77 ± 286.88                                             | 1557.18 ± 206.74                              |
+        | InvertedDoublePendulum-v4 | 5644.00 ± 353.46                                             | 296.97 ± 15.95                                |
+        | Reacher-v4                | -4.67 ± 0.88                                                 | -66.35 ± 0.66                                 |
+        | Swimmer-v4                | 124.52 ± 22.10                                               | 117.82 ± 10.07                                |
+        | Pusher-v4                 | -30.62 ± 6.80                                                | -276.32 ± 26.99                               |
+
+        Learning curves:
+        ![](../rpo/mujoco_v4_part2_0_5.png)
+
+
+        <!-- Tracked experiments: <iframe loading="lazy" src="https://wandb.ai/openrlbenchmark/cleanrl/reports/RPO-alpha-0-5-on-Mujoco_v4-Part-2--VmlldzozMjU4MTM1" style="height:500px;width:100%"> -->
```
Is this duplicate info since we have the "Failure case of `rpo_alpha=0.5`:" down below?

vwxyzjn replied:
Oh nvm I see what you did there

vwxyzjn reviewed (commented):
Hey I fixed some formatting issues with the iframe. Please let me know if you run into additional issues.

masud99r commented:
>

masud99r commented:
Thanks for the fix. Were those tab/space issues or something different?

> Hey I fixed some formatting issues with the iframe. Please let me know if you run into additional issues.

I have added a warning section with the depreciation notice.

With that, I think I have addressed all the comments, but I might lose some in the thread! Could you please take a look and let me know what else to be done for this PR?

vwxyzjn reviewed (approved):
LGTM. Thanks for your patience and this awesome contribution. Please give it a final read at https://cleanrl-git-fork-masud99r-master-vwxyzjn.vercel.app/rl-algorithms/rpo/. I have added you as a collaborator of this repo, and feel free to go ahead and merge.

masud99r commented:
> LGTM. Thanks for your patience and this awesome contribution. Please give it a final read at https://cleanrl-git-fork-masud99r-master-vwxyzjn.vercel.app/rl-algorithms/rpo/. I have added you as a collaborator of this repo, and feel free to go ahead and merge.

It was my pleasure working on this. Thanks for your excellent guidance throughout the process. I have learned a lot.

I took a pass on the docs and fixed some typos. I am going to merge.
